src/wp-includes/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -140,20 +140,36 @@ public function config( string $store_namespace, array $config = array() ): arra
 	 * @since 6.5.0
```

Comment:
I suspect this is late for 6.5, in that case an annotation should probably be added here that empty state objects are not printed.

Reply:
Hmm... it would be safer to print them, but always using objects to mimic JS, don't you think?

Something like this:

```php
function convert_empty_arrays_to_objects($data) {
    if (is_array($data)) {
        if (empty($data)) {
            return new stdClass();
        }
        
        foreach ($data as $key => $value) {
            $data[$key] = convert_empty_arrays_to_objects($value);
        }
    }
    
    return $data;
}
```

Reply:
That's an alternative. Why would it be safer? The client will use `{}` by default if no server provided state or config are provided, won't it?

Reply:
[The relevant client code is here:](https://github.com/WordPress/gutenberg/blob/2515c834ef4844ce63b8a9baca43e83a39a58784/packages/interactivity/src/store.ts#L313-L325)

```js
export const parseInitialData = ( dom = document ) => {
        const storeTag = dom.querySelector(
                `script[type="application/json"]#wp-interactivity-data`
        );
        if ( storeTag?.textContent ) {
                try {
                        return JSON.parse( storeTag.textContent );
                } catch ( e ) {
                        // Do nothing.
                }
        }
        return {};
};```

Reply:
🤔 This can actually change behavior right now if we omit stores. Subscriptions are slightly buggy if the store is not initialized before hydration. That's an edge case I hope to address in https://github.com/WordPress/gutenberg/pull/59842.

If that PR lands with the proposed approach, it seems preferable to avoid adding unnecessary bytes to the page to declare empty stores and configs.

Reply:
I think the fix is good. I would only prevent rendering an empty array when there's neither state nor config to include in the `<script>` tag. In addition, it would also save some bytes. 🙂

For arrays inside the state/config, that's something developers could handle on their own.

In any case, this fix is not critical, am I right? I mean, this is the function that initializes the store:

https://github.com/WordPress/gutenberg/blob/b06fcb1a7d4b3ae3b183625d7a1ae9a191e17ff7/packages/interactivity/src/store.ts#L327-L330

If an empty array were present, it would destructure the `state` and `config` properties, obtaining `undefined` for both.

Reply:
Disregard my previous comment. It should be up to the developer to indicate what will end up as an array and what will end up as an object:

```php
wp_interactivity_state( 'myPlugin', array(
  'array' => array(),
  'obj'   => (object) array(),
));
```

Reply:
> In any case, this fix is not critical, am I right? I mean, this is the function that initializes the store

Ah, I hadn't looked at that code. Here it is:

```js
export const populateInitialData = ( data?: {
        state?: Record< string, unknown >;
        config?: Record< string, unknown >;
} ) => {
        if ( isObject( data?.state ) ) {
                Object.entries( data.state ).forEach( ( [ namespace, state ] ) => {
                        store( namespace, { state }, { lock: universalUnlock } );
                } );
        }
        if ( isObject( data?.config ) ) {
                Object.entries( data.config ).forEach( ( [ namespace, config ] ) => {
                        storeConfigs.set( namespace, config );
                } );
        }
};
```

If the server provided initial data state is not an object then _the store is not created at all_. Again, this is related to https://github.com/WordPress/gutenberg/pull/59842 which will ensure that stores exist. I'd also consider adding a store to the client if invalid data was provided (I can add that to https://github.com/WordPress/gutenberg/pull/59842).

If https://github.com/WordPress/gutenberg/pull/59842 lands, this fix is entirely an optimization.

Reply:
My previous comment is wrong. This is still an important fix. Why? Because `state` is always expected to be an object. If we review the client's behavior, it parses the server provided data which has this form:

```jsonc
{
  state: {
    [namespace]: stateValueForNamespace,
    // …etc
  },
  config: {
    [config]: configValueForNamespace,
    // …etc
  },
}
```

But the client only looks at the top-level state and config object and skips store creation (or config set) if that level is not a plain object. Otherwise, it happily sets whatever is provided. Therefore, when the server does send arrays as state objects the client is not well behaved:

```json
{ "state": {
  "badState": [],
  "okState": {},
} }
```

There don't seem to be other plain object checks in the store function on the client.

<details>
<summary>The entire code block</summary>

```ts
export const parseInitialData = ( dom = document ) => {
	const storeTag = dom.querySelector(
		`script[type="application/json"]#wp-interactivity-data`
	);
	if ( storeTag?.textContent ) {
		try {
			return JSON.parse( storeTag.textContent );
		} catch ( e ) {
			// Do nothing.
		}
	}
	return {};
};

export const populateInitialData = ( data?: {
	state?: Record< string, unknown >;
	config?: Record< string, unknown >;
} ) => {
	if ( isObject( data?.state ) ) {
		Object.entries( data.state ).forEach( ( [ namespace, state ] ) => {
			store( namespace, { state }, { lock: universalUnlock } );
		} );
	}
	if ( isObject( data?.config ) ) {
		Object.entries( data.config ).forEach( ( [ namespace, config ] ) => {
			storeConfigs.set( namespace, config );
		} );
	}
};

// Parse and populate the initial state and config.
const data = parseInitialData();
populateInitialData( data );```

</details>

Reply:
Oh, right. Yup, this fix seems relevant. 😅